src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy

```diff
@@ -214,6 +214,46 @@ class GraphQLSchemaTest extends Specification {
         dogType.getName() == "Dog"
     }
 
+    def "issue with type references when original type is transformed away"() {
+        def sdl = '''
+            type Query {
+              # The b fields leads to the addition of the B type (actual definition)
+              b: B
+              # When we filter out the `b` field, we can still access B through A
+              # however they are GraphQLTypeReferences and not an actual GraphQL Object
+              a: A
+            } 
```

Comment:
A couple of comments - you have a transient `Query.a` -> `type A`  -> `A.a` -> `type A` - while the schema is being built type references are indeed used HOWEVER once the schema is fully built (a late step) then all type references are removed and replaced with actual object pointers.   This means that the runtime types are in fact NOT Immutable but are mutable by design.

So you mentioned GraphQLTypeReferences but these would not exist by the time `def schema = TestUtil.schema(sdl)` returned

Reply:
Yeah the problem is that during `build` and `buildImpl` the schema partially built at first, and the transformed roots do indeed contain GraphQLTypeReferences.

I just saw you confirmed the bug with `transformSchema` as well, I should have mentioned this hits the same problem as well yeah 👍